tiledb/sm/serialization/array.cc

```diff
@@ -358,6 +365,246 @@ Status array_open_from_capnp(
+            "[fragments_timestamps_serialize] Unknown serialization type "
+            "passed");
+      }
+    }
+
+  } catch (kj::Exception& e) {
+    throw ArraySerializationException(
+        "[fragments_timestamps_serialize] kj::Exception: " +
+        std::string(e.getDescription().cStr()));
+  } catch (std::exception& e) {
+    throw ArraySerializationException(
+        "[fragments_timestamps_serialize] exception " + std::string(e.what()));
+  }
+}
+
+std::tuple<uint64_t, uint64_t> fragments_timestamps_deserialize(
+    SerializationType serialize_type, const Buffer& serialized_buffer) {
+  try {
+    switch (serialize_type) {
+      case SerializationType::JSON: {
+        ::capnp::JsonCodec json;
+        ::capnp::MallocMessageBuilder message_builder;
+        auto builder =
+            message_builder.initRoot<capnp::ArrayFragmentsTimestamps>();
+        json.decode(
+            kj::StringPtr(static_cast<const char*>(serialized_buffer.data())),
+            builder);
+        auto reader = builder.asReader();
+        // Deserialize
+        return fragments_timestamps_from_capnp(reader);
+      }
+      case SerializationType::CAPNP: {
+        const auto mBytes =
+            reinterpret_cast<const kj::byte*>(serialized_buffer.data());
+        ::capnp::FlatArrayMessageReader msg_reader(kj::arrayPtr(
+            reinterpret_cast<const ::capnp::word*>(mBytes),
+            serialized_buffer.size() / sizeof(::capnp::word)));
+        auto reader = msg_reader.getRoot<capnp::ArrayFragmentsTimestamps>();
+        // Deserialize
+        return fragments_timestamps_from_capnp(reader);
+      }
+      default: {
+        throw ArraySerializationException(
+            "[fragments_timestamps_deserialize] "
+            "Unknown serialization type passed");
+      }
+    }
+  } catch (kj::Exception& e) {
+    throw ArraySerializationException(
+        "[fragments_timestamps_deserialize] kj::Exception: " +
+        std::string(e.getDescription().cStr()));
+  } catch (std::exception& e) {
+    throw ArraySerializationException(
+        "[fragments_timestamps_deserialize] exception " +
+        std::string(e.what()));
+  }
+}
+
+void fragments_list_to_capnp(
+    const std::vector<URI>& fragments,
+    capnp::ArrayFragmentsList::Builder* array_fragments_list_builder) {
+  auto entries_builder =
+      array_fragments_list_builder->initEntries(fragments.size());
+  for (size_t i = 0; i < fragments.size(); i++) {
+    const auto& relative_uri = serialize_array_uri_to_relative(fragments[i]);
+    entries_builder.set(i, relative_uri);
+  }
+}
+
+std::vector<URI> fragments_list_from_capnp(
+    const capnp::ArrayFragmentsList::Reader& array_fragments_list_reader,
+    const URI& array_uri) {
+  if (array_fragments_list_reader.hasEntries()) {
+    std::vector<URI> fragments;
+    fragments.reserve(array_fragments_list_reader.getEntries().size());
+    for (auto entry : array_fragments_list_reader.getEntries()) {
```

Comment:
get an `auto get_entries_reader = array_fragments_list_reader.getEntries();` before the loop, I've recently learnt capnp counts each of those calls to traversal limit.